src/vs/code/electron-main/backup.ts

```diff
@@ -0,0 +1,88 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import * as fs from 'original-fs';
+import * as path from 'path';
+import { createDecorator } from 'vs/platform/instantiation/common/instantiation';
+import { IEnvironmentService } from 'vs/platform/environment/common/environment';
+
+export const IBackupService = createDecorator<IBackupService>('backupService');
+
+export interface IBackupService {
+	getBackupWorkspaces(): string[];
+	clearBackupWorkspaces(): void;
+	pushBackupWorkspaces(workspaces: string[]): void;
+	getBackupFiles(workspace: string): string[];
+}
+
+interface IBackupFormat {
+	folderWorkspaces?: {
+		[workspacePath: string]: string[]
+	};
+}
+
+export class BackupService implements IBackupService {
+
+	private filePath: string;
+	private fileContent: IBackupFormat;
+
+	constructor(
+		@IEnvironmentService private environmentService: IEnvironmentService
+	) {
+		this.filePath = path.join(environmentService.userDataPath, 'Backups', 'workspaces.json');
```

Comment:
I like it https://github.com/Microsoft/vscode/pull/12400/commits/5f230ed17cadf8122ec268bc2d953247dc988305